### ISSUE TITLE
Unable to use docker_service because ansible-docker installs docker-py

### ISSUE THREAD

**Issue by amarandon** [nickjj/ansible-docker#40]
I use the `docker_service` Ansible role to start services based on a `docker-compose.yml` file. This role depends on the `docker-compose` Python module which itself depends on the `docker` Python module. But ansible-docker installs the `docker-py` Python module and it's not possible to install both `docker` and `docker-py` Python modules at the same time.

More info: https://docs.ansible.com/ansible/latest/modules/docker_service_module.html#requirements

I'm going to try changing my copy of docker-ansible so that it installs `docker` instead of `docker-py` and see how it goes.

**Comment by nickjj** [comment]
Hi,

This role installs Docker using Docker's official Debian / Ubuntu package. Compose is also copied over directly as a binary (not installed through Python).

This role shouldn't be installing `docker-py`. I would test it on a fresh system.

**Comment by amarandon** [comment]
Hi Nick, more info below!

> This role installs Docker using Docker's official Debian / Ubuntu package. Compose is also copied over directly as a binary (not installed through Python)

Indeed I'm talking about Python modules here, not the Go binaries. `docker_service` uses the `docker-compose` Python module which as I understand it is an alternative, pure Python implementation of docker-compose : https://pypi.org/project/docker-compose/

> This role shouldn't be installing docker-py

I use ansible-docker to login to Docker registries and this what seems to trigger the installation of docker-py:
https://github.com/nickjj/ansible-docker/blob/master/tasks/main.yml#L114

**Comment by nickjj** [comment]
Oh yeah, that's because the Ansible `docker_login` module requires the `docker-py` package to be installed. Its dependencies are listed at https://docs.ansible.com/ansible/2.5/modules/docker_login_module.html.

I'm having trouble understanding the real issue though.

- The login module installs docker-py
- The service module requires docker-compose
- This role roles docker-py (if you're using login)
- This role installs Docker Compose

All of the requirements to run both the login and server modules should be there?

**Comment by nickjj** [comment]
When you say this:

_"I've tried to amend ansible-docker so that it installs docker instead of docker-py and it seems to be working for me so far."_

Are you talking about pip installing `docker` instead of `docker-py` to meet the `docker_login` module's requirements? When you do that, both `docker_login` and `docker_service` works?

I never installed Compose with pip, mainly because I don't use `docker_service` or any Ansible Docker modules (except for logging into a registry).

The registry management could be done without `docker_login` but it's kind of annoying to make it idempotent.

I don't mind switching to to using `docker` instead of `docker-py` as long as `docker_login` still works. If you want to set up a test playbook where you use both login and service and it demonstrates it working, that would be a good start.

**Comment by amarandon** [comment]
> Are you talking about pip installing docker instead of docker-py to meet the docker_login module's requirements? When you do that, both docker_login and docker_service works?

Yes this is what I meant!

> I don't mind switching to using docker instead of docker-py as long as docker_login still works

Yes in my experience, it is the case. In fact, until now I have been using `docker_login` directly while having installed docker-compose with pip (which installed the `docker` Python module). 

Now that I found out the `docker__registries` option of `ansible-docker`, I thought it would be good to switch to it as it makes things slightly neater.

I'll look into making a playbook demonstrating the use case.

**Comment by amarandon** [comment]
Here is the playbook
```yaml
---

- hosts: all
  become: true
  tasks:
    - name: Install required packages
      apt:
        state: present
        update_cache: true
        name:
          - python-pip
          - python-setuptools

- hosts: all
  become: true
  vars:
    docker__registries:
      - registry_url: registry.gitlab.com
        unsername: gitlab.com
        password: "{{ password }}"
  roles:
    - role: nickjj.docker
  tasks:
    - name: Install docker-compose python module
      pip:
        name: docker-compose
    - name: Copy docker-compose.yml
      copy:
        src: docker-compose.yml
        dest: docker-compose.yml
    - name: Start nginx
      docker_service:
        state: present
        project_src: .

```

And the `docker-compose.yml`:
```yaml
version: "3"

services:
  nginx:
    image: nginx:1.13.6-alpine
    ports:
      - "80:80"
```

This gives me this error:
```yaml
TASK [Start nginx] ********************************************************************
fatal: [debian@XX.XX.XXX.XX]: FAILED! => {"changed": false, "msg": "Cannot have
both the docker-py and docker python modules installed together as they use the
same namespace and cause a corrupt installation. Please uninstall both packages,
and re-install only the docker-py or docker python module. It is recommended to
install the docker module if no support for Python 2.6 is required. Please note that
simply uninstalling one of the modules can leave the other module in a broken state."}
        to retry, use: --limit @/home/al/essai/ansible/playbook.retry
```

Then I change `nickjj.docker/tasks/main.yml` so that it installs `docker` instead of `docker-py`:
```yaml
- name: Install docker for managing Docker login credentials
  pip:
    name: "docker"
  when: docker__registries
```
Then I reset my VM and run ansible-playbook again and everything works fine. Docker login succeeds and nginx works.

**Comment by nickjj** [comment]
Great, thanks. I will test this in a day or 2 and get back to you.

Then you could either make a PR with your change, or I can do it (up to you).

**Comment by nickjj** [comment]
If you decide you want to do the PR, you can do that before then if you want and I'll merge it after I test it here.

**Comment by nickjj** [comment]
I had to act fast because Docker 18.09 shipped and they changed something big when it comes to their systemd unit file. Long story short, this role was broken by default with 18.09.

v1.5.0 is now pushed which fixes the issue in the following way:

1. This role now apt installs `python-pip` by default.
2. This role now pip installs `docker` always.

The reason I went for "always" instead of just when at least 1 Docker login was set was because you might want to use `docker_service` without logging in, in which case, the pip docker package needs to be installed.

In the future, perhaps the pip install step could be turned into a variable which is also tied into installing python-pip in the case where you don't use Docker login or the service module and want to avoid installing both of those packages.
